types/underscore/index.d.ts

```diff
@@ -3948,6 +3948,50 @@ declare namespace _ {
          */
         isUndefined(object: any): object is undefined;
 
+        /**
+         * Returns true if `object` is a Set.
+         * @param object The object to check.
+         * @returns True if `object` is a Set, otherwise false.
+         */
+        isSet(object: any): object is Set<any>;
+
+        /**
+         * Returns true if `object` is a WeakSet.
+         * @param object The object to check.
+         * @returns True if `object` is a WeakSet, otherwise false.
+         */
+        isWeakSet(object: any): object is WeakSet<any>;
+
+        /**
+         * Returns true if `object` is a Map.
+         * @param object The object to check.
+         * @returns True if `object` is a Map, otherwise false.
+         */
+        isMap(object: any): object is Map<any, any>;
+
+        /**
+         * Returns true if `object` is a WeakMap.
+         * @param object The object to check.
+         * @returns True if `object` is a WeakMap, otherwise false.
+         */
+        isWeakMap(object: any): object is WeakMap<any, any>;
+
+        /**
+         * Ensures that path is an array.
+         * @param path
+         *  1. If path is a string, it is wrapped in a single-element array;
+         *  2. if it is an array already, it is returned unmodified.
+         */
+        toPath<P>(path: P): (
+            // eslint-disable-next-line @definitelytyped/no-single-element-tuple-type
+            P extends ReadonlyArray<string | number> ? [...P]
```

Comment:
Why the separate specialization for `ReadonlyArray`? Does it not extend `Array`?

Reply:
Also, why not just return `P`?

Reply:
> Why the separate specialization for ReadonlyArray? Does it not extend Array?

Actually `extends ReadonlyArray` is enough. I have removed the other cases in latest commit.

> Also, why not just return P?

If `extends ReadonlyArray` and return `P`, the return type will continue to be read-only, which may not be expected by the consumer. 

Ideally after they get a path array back, they can do push/pop or whatever they want on the array. In this case, returning `[...P]` can "retain" the original array shape, and make the return no longer read-only.

You can see that in `underscore.get` typedef, same trick is also used.

Reply:
The thing is, if `path` is already array-like, `toPath` returns the original value rather than a copy. So if the original value is readonly, so should the return value, I think. See [the source](http://underscorejs.org/docs/modules/toPath.html).

Reply:
Yes, you are right. I have just updated that part.